### PR TITLE
[DSF-7703] ADA Timeout - logout announcement

### PR DESCRIPTION
**Pull request by emiclark**
### Changelog
Added a timer to announce a warning 30 seconds before logout

### CODE REVIEW

File: Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift [hunk 1]
```diff
@@ -51,19 +56,30 @@ extension IdleTimer {
 
 extension IdleTimer {
     final class WindowContainer {
-        private let timer: InternalTimer
+        private let timer1: InternalTimer
+        private let timer2: InternalTimer
```
Comment (adeca) [inline Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift]: I suggest renaming `timer1` to `timer`, and `timer2` to `warningTimer` or `alertTimer`.

File: Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+InternalTimer.swift [hunk 2]
```diff
@@ -59,6 +64,12 @@ extension IdleTimer {
             })
         }
 
+        private func handleVOAnnouncement() {
```
Comment (adeca) [inline Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+InternalTimer.swift]: Posting the warning announcement should not be part of this class. This is a generic timer which just calls an `onTimeout` block.

**Review by adeca — CHANGES_REQUESTED** [review]

File: Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer.swift [hunk 3]
```diff
@@ -19,7 +19,7 @@ extension IdleTimer {
     ///
     /// You can observe the timeout event using
     /// `UIApplication.didTimeOutUserInteractionNotification`.
-    public static func setUserInteractionTimeout(duration: TimeInterval? = nil, for window: UIWindow?) {
+        public static func setUserInteractionTimeout(duration: TimeInterval? = nil, for window: UIWindow?) {
```
Comment (adeca) [inline Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer.swift]: Lets revert this

File: Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift [hunk 4]
```diff
@@ -25,6 +30,12 @@ extension NotificationCenter.Event {
     public func applicationDidTimeOutUserInteraction(_ callback: @escaping () -> Void) -> NSObjectProtocol {
         observe(UIApplication.didTimeOutUserInteractionNotification, callback)
     }
+
+    /// - See: `IdleTimer.setUserInteractionTimeout(duration:for:)`
+    @discardableResult
+    public func applicationWillTimeOutIdleTimerNotification(_ callback: @escaping () -> Void) -> NSObjectProtocol {
```
Comment (adeca) [inline Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift]:
```suggestion
    public func applicationWillTimeOutUserInteraction(_ callback: @escaping () -> Void) -> NSObjectProtocol {
```
This makes it consistent with the `applicationDidTimeOutUserInteraction` method.

File: Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift [hunk 5]
```diff
@@ -15,6 +15,11 @@ extension UIApplication {
     public static var didTimeOutUserInteractionNotification: Notification.Name {
         .init(#function)
     }
+
+    /// - See: `IdleTimer.willTimeoutIdleTimer(duration:for:)`
+    public static var willTimeOutIdleTimerNotification: Notification.Name {
```
Comment (adeca) [inline Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift]:
```suggestion
    public static var willTimeOutUserInteractionNotification: Notification.Name {
```
This makes it consistent with the `didTimeOutUserInteractionNotification` property.

**Review by adeca — CHANGES_REQUESTED** [review]

**Review by adeca — APPROVED** [review]

File: Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+InternalTimer.swift [hunk 6]
```diff
@@ -12,7 +12,6 @@ extension IdleTimer {
         private var uptime = MonotonicClock.Uptime()
         private var timer: MonotonicClock.Timer?
         private let onTimeout: () -> Void
-
```
Comment (mstana) [inline Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+InternalTimer.swift]: revert this change please

File: Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift [hunk 7]
```diff
@@ -52,18 +63,27 @@ extension IdleTimer {
 extension IdleTimer {
     final class WindowContainer {
         private let timer: InternalTimer
+        private let warningTimer: InternalTimer
+        private let warningTime: TimeInterval = 30 // seconds before voiceover announces the logout warning
```
Comment (mstana) [inline Sources/Swift/Components/Dates and Timers/IdleTimer/IdleTimer+Gesture.swift:70]:
```suggestion
        /// The timeout in seconds specifies the duration before the main `timer` notification is posted and will
        /// result in posting separate warning notification.
        private let warningTime: TimeInterval = 30
```

**Comment by mstana** [comment]
We will handle this by https://github.com/claritymoney/xcore/pull/9

**Pull request closed without merging**